nodes/image_sub_matcher.py

```diff
@@ -0,0 +1,218 @@
+
+	def append_teacher_full_image(self, msg):
+		self.full_images.append(image_processing.msg_to_image(msg))
+
+	def append_teacher_norm_image(self, msg):
+		self.norm_images.append(image_processing.msg_to_image(msg))
+
+	# Do Not need to load images from directory for Multi-Robot System
+	#def load_image_left_right(self, file_left, file_right):
+	#	image_left = image_processing.grayscale(cv2.imread(file_left))
+	#	image_right = image_processing.grayscale(cv2.imread(file_right))
+	#	image_both, _ = image_processing.rectify_stitch_stereo_image(image_left, image_right, self.cam_left_calibration, self.cam_right_calibration)
+	#	return image_processing.patch_normalise_pad(cv2.resize(image_both,  self.resize[::-1], interpolation=cv2.INTER_AREA), self.patch_size)
+
+	#def load_images_left_right(self, image_files):
+	#	return [self.load_image_left_right(*image_pair) for image_pair in image_files]
+
+	#def load_images_resize_norm(self, image_files):
+	#	images = [cv2.imread(image_file, cv2.IMREAD_GRAYSCALE) for image_file in image_files]
+	#	return [image_processing.patch_normalise_pad(cv2.resize(image,  self.resize[::-1], interpolation=cv2.INTER_AREA), self.patch_size) for image in images]
+
+	#def load_images_raw(self, image_files):
+		# these images are already patch normalised (should be -1 -> 1), so convert them from uint8 (0 -> 255)
+	#	return [-1 + 2./255.*cv2.imread(image_file, cv2.IMREAD_GRAYSCALE) for image_file in image_files]
+
+	#def weight_images_depth(self, images, image_files):
+	#	depth_images = [np.load(f[:-4]+'_disp.npy').squeeze() for f in image_files]
+	#	depth_images = [np.clip(cv2.resize(image, self.resize[::-1], interpolation=cv2.INTER_AREA), 0, 1) for image in depth_images]
+		# depth_images = [np.maximum(image, image > 0.5) for image in depth_images]
+	#	return [image * depth for image, depth in zip(images, depth_images)]
+
+	#def weight_images_middle(self, images):
+	#	horizontal_weighting = np.clip(1.0 - 1.0*(np.abs(np.arange(self.resize[1]) - (self.resize[1]/2.)).reshape((1,-1)) / (self.resize[1]/2.))**2, 0, 1)
+	#	vertical_weighting = np.clip(1.0 + 0*np.linspace(0,1,self.resize[0]).reshape(-1,1)**0.5, 0, 1)
+	#	image_weighting = vertical_weighting * horizontal_weighting
+	#	return [image * image_weighting for image in images]
+
+	#def read_file(self, filename):
+	#	with open(filename, 'r') as f:
+	#		data = f.read()
+	#	return data
+
+	def clamp_search_range_to_bounds(self, index, half_search_range):
+		start_range = max(0, index - half_search_range)
+		end_range = min(len(self.images), index + half_search_range + 1)
+		return start_range, end_range
+
+	def match_image(self, request):
+		# Added in to wait till there is enough data before proceeding to follow
+		self.images = self.norm_images
+		while(len(self.images) - request.imageIndex.data < self.frames_delay):
+			self.images == self.norm_images
```

Comment:
So this is the main part that causes the localiser on the student to wait until the teacher has collected enough images?

Reply:
Yes, that is correct.
It basically just wait till the specified numbers of frames received.

Reply:
The line `self.images == self.norm_images` specifically doesn't do anything though? Is that intended?

Reply:
No that was not intended, I should've just use pass.
I will change that, I am unsure why I did it there.

Reply:
have fixed up the code and removed unnecessary codes